### PR TITLE
--emitDeclarationsOnly flag to enable declarations only output

### LINKED ISSUE #15158: Suggestion: a --noEmitJS compiler option
## The problem

It is sometimes useful to set different compiler options for `.js` files and `.d.ts` files. For example, I often want to set `--removeComments` for JavaScript outputs but not for declaration files. This can currently be achieved by building a project with `--declaration` and `--removeComments` enabled, removing emitted `.js` files, and then building the project without those options:

```sh
tsc --declaration --removeComments && rimraf "dest/**/*.js" && tsc
```

This is convoluted and wasteful. Of course, there are always tools like Gulp, but it would be nice to have a solution using just the `tsc` command.

## A solution: a --noEmitJS compiler option

I would like to see a `--noEmitJS` compiler option. Like `--noEmit`, it would suppress emitted JavaScript files, but unlike `--noEmit`, it would still emit sourcemaps and declaration files if the relevant compiler options are enabled. The above could be achieved like so:

```sh
tsc && tsc --declaration --removeComents --noEmitJS
```

This is nicer. It would also cover any other scenarios where different options are required for either declarations or sourcemaps.

### PR DESCRIPTION
**Pull request by nojvek**
- [x] Fixes #15158 and #1866. Part of fix for #7546
- [x] Code is up-to-date with the `master` branch
- [x] You've successfully run `jake runtests` locally
- [x] You've signed the CLA
- [x] There are new or updated unit tests validating the change

### CODE REVIEW

File: src/compiler/program.ts [hunk 1]
```diff
@@ -2186,6 +2186,14 @@ namespace ts {
...
             }
 
+            if (options.noEmit && options.noEmitJs) {
+                createDiagnosticForOptionName(Diagnostics.Option_0_cannot_be_specified_with_option_1, "noEmit", "noEmitJs");
+            }
+
+            if (options.noEmitJs && !options.declaration) {
+                programDiagnostics.add(createCompilerDiagnostic(Diagnostics.Option_0_cannot_be_specified_without_specifying_option_1, "noEmitJs", "declaration"));
+            }
+
```
Comment (sheetalkamat) [inline src/compiler/program.ts]: you would also want to skip ```verifyEmitFilePath(emitFileNames.jsFilePath, emitFilesSeen);``` down below when ```noEmitJs``` is ```true```

**Comment by nojvek** [comment]
Kind ping regarding this @sheetalkamat 

Should I be adding any specific devs from typescript team to this PR? What's the usual process before a PR lands in the repo ?

File: src/compiler/commandLineParser.ts [hunk 2]
```diff
@@ -231,6 +231,13 @@ namespace ts {
             category: Diagnostics.Basic_Options,
             description: Diagnostics.Do_not_emit_outputs,
         },
+        {
+            name: "noEmitJs",
```
Comment (mhegazy) [inline src/compiler/commandLineParser.ts]: I think the name should be `--emitDeclarationsOnly` instead.

File: src/compiler/commandLineParser.ts [hunk 3]
```diff
@@ -231,6 +231,13 @@ namespace ts {
             category: Diagnostics.Basic_Options,
             description: Diagnostics.Do_not_emit_outputs,
         },
+        {
+            name: "noEmitJs",
+            type: "boolean",
+            showInSimplifiedHelpView: true,
+            category: Diagnostics.Basic_Options,
```
Comment (mhegazy) [inline src/compiler/commandLineParser.ts]: i would say this belongs into `Advanced_Options`

File: src/compiler/diagnosticMessages.json [hunk 4]
```diff
@@ -2776,6 +2776,10 @@
         "category": "Message",
         "code": 6013
     },
+    "Do not emit js outputs.": {
```
Comment (mhegazy) [inline src/compiler/diagnosticMessages.json]: `Only emit declaration files.`

File: src/server/server.ts [hunk 5]
```diff
@@ -980,7 +980,7 @@ namespace ts.server {
     };
 
     const ioSession = new IOSession(options);
-    process.on("uncaughtException", err => {
+    process.on("uncaughtException", (err: any) => {
```
Comment (mhegazy) [inline src/server/server.ts]: why?

Comment (nojvek) [reply src/server/server.ts]: Because it was failing compilation when doing a “gulp build” with an error.

Comment (mhegazy) [reply src/server/server.ts]: should work now. that was a @types/node issue.

**Comment by niieani** [comment]
@nojvek the Puppeteer community would love some progress on this! ❤️

**Comment by nojvek** [comment]
On it.

File: src/compiler/commandLineParser.ts (hunk 3 above)
Comment (nojvek) [reply src/compiler/commandLineParser.ts]: AdvancedOptions won't show up in `--help` and this is something that will definitely be very useful with allowJs and checkJs projects to generate their typings. Basically enabling thousands of npm js repos to automatically generate typings if they've annotated code with jsdoc.

Although the PR doesn't include full understanding of prototype based functions. Does it work with you if we put this in experimental mode first ?

Comment (nojvek) [reply src/compiler/commandLineParser.ts]: Advanced_options don't show up as part of `--help` or in default tsc --init. Doesn't it make sense that the option showed up there? I plan to enable allowJs + emitDeclarationsOnly in another PR which would help with lots of npm projects to auto-generate jsdoc to .d.ts declarations

**Comment by nojvek** [comment]
@mhegazy want to take another pass ?

@niieani - Unfortunately this PR doesn't enable emitting declarations from a JS project.  I'll start work on it as soon as this PR lands.

**Comment by mhegazy** [comment]
@DanielRosenwasser and @sheetalkamat any concerns?

**Comment by niieani** [comment]
Thank you @nojvek! Looking forward to the next PR 😊!

**Comment by nojvek** [comment]
@mhegazy @sheetalkamat 

```
// @declaration: true
// @emitDeclarationsOnly: true
```

For consistency with `declaration: true` flag, does it make sense to call this new flag `emitDeclarationOnly` ?

```
        declaration?: boolean;
        emitDeclarationsOnly?: boolean;
        declarationDir?: string;
```

It seems the pattern is to use singular `*declaration*`.

What do you think ? 

I can send a rename PR

**Comment by mhegazy** [comment]
humm.. good point.. we can change it to `EmitDeclarationOnly`

**Comment by nojvek** [comment]
K will send a rename PR

**Comment by mhegazy** [comment]
rename PR up in https://github.com/Microsoft/TypeScript/pull/21651

**Pull request merged**